[DOC] Add ZJIT glossary table (#13860)
ruby/ruby: doc/zjit.md

```
make test-all TESTS="test/ruby/test_zjit.rb -n TestZJIT#test_putobject"
```

## ZJIT Glossary

This glossary contains terms that are helpful for understanding ZJIT.

Please note that some terms may appear in CRuby internals too but with different meanings.

| Term | Definition |
| ---  | -----------|

> | HIR | High-level Intermediate Representation. High-level intermediary representation in static single-assignment (SSA) form |

Reviewer: ```suggestion
| HIR | High-level Intermediate Representation. High-level (Ruby semantics) graph representation in static single-assignment (SSA) form |
```

| LIR | Low-level Intermediate Representation. Low-level IR used in the backend for assembly generation |
| SSA | Static Single Assignment. A form where each variable is assigned exactly once |
| `opnd` | Operand. An operand to an IR instruction (can be register, memory, immediate, etc.) |
| `dst` | Destination. The output operand of an instruction where the result is stored |
| VReg | Virtual Register. A virtual register that gets lowered to physical register or memory |
| `insn_id` | Instruction ID. An index of an instruction in a function |
| `block_id` | The index of a basic block, which effectively acts like a pointer |
| `branch` | Control flow edge between basic blocks in the compiled code |
| `cb` | Code Block. Memory region for generated machine code |
| `entry` | The starting address of compiled code for an ISEQ |

> | Patch Point | Location in generated code that can be modified later for invalidation or linking |

Reviewer: ```suggestion
| Patch Point | Location in generated code that can be modified later in case assumptions get invalidated |
```

| Frame State | Captured state of the Ruby stack frame at a specific point for deoptimization |

> | Guard | A runtime check that ensures assumptions are still valid |

Reviewer: ```suggestion
| Guard | A run-time check that ensures assumptions are still valid |
```
